cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactional.java

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright © 2016 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.app.runtime.spark;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.data.DatasetContext;
+import co.cask.cdap.api.data.DatasetInstantiationException;
+import co.cask.cdap.api.dataset.Dataset;
+import co.cask.cdap.data2.dataset2.DynamicDatasetCache;
+import co.cask.cdap.data2.transaction.Transactions;
+import co.cask.tephra.Transaction;
+import co.cask.tephra.TransactionAware;
+import co.cask.tephra.TransactionFailureException;
+import co.cask.tephra.TransactionSystemClient;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.Iterables;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link Transactional} for managing transactions for Spark job execution.
+ */
+final class SparkTransactional implements Transactional {
+
+  /**
+   * Property key for storing the key to lookup active transaction
+   */
+  static final String ACTIVE_TRANSACTION_KEY = "cdap.spark.active.transaction";
+
+  private static final Logger LOG = LoggerFactory.getLogger(SparkTransactional.class);
+
+  private final ThreadLocal<TransactionalDatasetContext> activeDatasetContext =
+    new ThreadLocal<TransactionalDatasetContext>() {
+      @Override
+      public void set(TransactionalDatasetContext value) {
+        String txKey = Long.toString(value.getTransaction().getWritePointer());
+        if (SparkContextCache.setLocalProperty(ACTIVE_TRANSACTION_KEY, txKey)) {
```

Comment:
It will be false when there is no SparkContext. E.g. when user code look like this:

```java
void run(SparkExeuctionContext sec) {
  sec.execute(new TxRunnable() {
    // In here, there is no SparkContext when creating a new transaction.
  });

  // The SparkContext create after something is executed in a transaction.
  // E.g. user can read a dataset of config and use it to create the SparkConf that passes to the SparkContext.
  JavaSparkContext jsc = new JavaSparkContext(...);
}
```